File: app/src/main/kotlin/br/com/victorwads/goldenunicorn/data/models/CreditRegistry.kt
package br.com.victorwads.goldenunicorn.data.models

import com.google.firebase.firestore.DocumentId
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.PropertyName
import java.util.Date

data class CreditRegistry(
    @DocumentId val id: String? = null,
    @PropertyName("cardId") val bankId: String,
    @PropertyName("invoice") val invoice: String,

    @PropertyName("name") val name: String,
    @PropertyName("type") val type: Type = Type.DEFAULT,
    @PropertyName("date") val date: Date,
    @PropertyName("metadata") val metadata: Metadata? = null,

    @PropertyName("createdAt")
    val createdAt: FieldValue? = FieldValue.serverTimestamp(),
) {
    @PropertyName("updatedAt")
    val updatedAt: FieldValue = FieldValue.serverTimestamp()

    enum class Type {
        DEFAULT, FIXED
    }

    data class Metadata(
        @PropertyName("installmentBillId") val installmentBillId: String? = null,
    )
}

File: iOS/Sources/Data/models/CreditRegistry.swift

import Foundation
import FirebaseFirestore

struct CreditRegistry: Codable, Identifiable {
    @DocumentID var id: String? = nil
    var cardId: String
    var categoryId: String
    var invoice: String

    var name: String = ""
    var amount: Double = 0
    var date: Date = Date()
    var type: Type = .default

    var metadata: Metadata = Metadata()
    var createdAt: Date = Date()
    var updatedAt: Date = Date()
    
    enum CodingKeys: String, CodingKey {
        case id
        case cardId = "cardId"
        case categoryId = "categoryId"
        case invoice = "invoice"

        case name = "name"
        case amount = "amount"
        case date = "date"
        case type = "type"
        case metadata = "metadata"

        case createdAt = "createdAt"
        case updatedAt = "updatedAt"
    }

    enum `Type`: String, Codable {
        case `default` = "DEFAULT"
        case fixed = "FIXED"
        case installment = "INSTALLMENT"
    }

    struct Metadata: Codable {
        var installmentBillId: String? = nil
        var fixedBillId: String? = nil

        enum CodingKeys: String, CodingKey {
            case installmentBillId = "installmentBillId"
            case fixedBillId = "fixedBillId"
        }
    }
}
